Objects/obmalloc.c

```diff
@@ -86,6 +92,135 @@ _PyMem_RawFree(void *Py_UNUSED(ctx), void *ptr)
 
 #ifdef WITH_MIMALLOC
 
+static void
+_PyMem_mi_page_clear_qsbr(mi_page_t *page)
+{
+#ifdef Py_GIL_DISABLED
+    // Clear the QSBR goal and remove the page from the QSBR linked list.
+    page->qsbr_goal = 0;
+    if (page->qsbr_node.next != NULL) {
+        llist_remove(&page->qsbr_node);
+    }
+#endif
+}
+
+// Is the empty page safe to free? It's safe if there was a QSBR goal set and
+// the goal has been reached.
+static bool
+_PyMem_mi_page_is_safe_to_free(mi_page_t *page)
+{
+    assert(mi_page_all_free(page));
+#ifdef Py_GIL_DISABLED
+    if (page->use_qsbr) {
+        if (page->qsbr_goal == 0) {
+            // No QSBR goal set, so we can't safely free the page yet.
+            return false;
+        }
+        _PyThreadStateImpl *tstate = (_PyThreadStateImpl *)_PyThreadState_GET();
+        if (tstate == NULL) {
+            return false;
+        }
+        if (!_Py_qbsr_goal_reached(tstate->qsbr, page->qsbr_goal)) {
+            return false;
+        }
+        _PyMem_mi_page_clear_qsbr(page);
```

Comment:
This side effect seems surprising given the name of the function...  It seems like it'd be better to call `_PyMem_mi_page_clear_qsbr` explicitly in `mi_segment_page_clear` and `_PyMem_mi_page_maybe_free`